README.md
# Requirements
To install requirements
```
pip install -r requirements.txt
```
Use Python version 3.7.4. 

# Training
## Training RenyiGAN-[&beta;<sub>1</sub>, &beta;<sub>2</sub>]
To train RenyiGAN-[&beta;<sub>1</sub>, &beta;<sub>2</sub>], run this command
```
python3 renyigan_varying_alpha.py
```
which will prompt you to input the seed, trial number, version, and subversion. 

The version and subversion determines the values of &beta;<sub>1</sub> and &beta;<sub>2</sub>
in RenyiGAN-[&beta;<sub>1</sub>, &beta;<sub>2</sub>] 
as well as whether or not you activate L1 normalization and/or gradient penalty. 
The table below details this information:

| Version | Subversion | [&beta;<sub>1</sub>, &beta;<sub>2</sub>] | Gradient penalty  | L1 normalization |
| :---: | :---: | :---: | :---: | :---: |
| 1 | 1 | [0, 0.9] | &#10006; | &#10006; |
| 1 | 2 | [0, 3.0] | &#10006; | &#10006; |
| 1 | 3 | [1.1, 4.0] | &#10006; | &#10006; |
| 2 | 1 | [0, 0.9] | &#10006; | &#10004; |
| 2 | 2 | [0, 3.0] | &#10006; | &#10004; |
| 2 | 3 | [1.1, 4.0] | &#10006; | &#10004; |
| 3 | 1 | [0, 0.9] | &#10004; | &#10006; |
| 3 | 2 | [0, 3.0] | &#10004; | &#10006; |
| 3 | 3 | [1.1, 4.0] | &#10004; | &#10006; |
| 4 | 1 | [0, 0.9] | &#10004; | &#10004; |
| 4 | 2 | [0, 3.0] | &#10004; | &#10004; |
| 4 | 3 | [1.1, 4.0] | &#10004; | &#10004; |


The code will automatically download the MNIST database using the `utils.py` file. 
It will also save the generated images as `.npy` files in `data/annealing/`. 
Note that the code only saves the checkpoints for the first trial in the `checkpoints` directory. 

## Training RenyiGAN-&alpha;
To train RenyiGAN-&alpha; (i.e. RenyiGAN for a fixed value &alpha; for all epochs during training), run this command:
```
python3 renyigan_static_alpha.py
``` 
which will prompt you to input the alpha value, trial number, version, and seed. 
We used seed 123, 5005, 1600, 199621, 60677, 20435, 15859, 33764, 79878, 
36123 for trials 1 to 10, respectively.
The version number determines whether or not you activate L1 normalization and/or gradient penalty.
The table below details this information:

| Version | Gradient Penalty | L1 normalization |
| :---: | :---: | :---: |
| 1 | &#10006; | &#10006; |
| 2 | &#10006; | &#10004; |
| 3 | &#10004; | &#10006; |
| 4 | &#10004; | &#10004; |

Note that if you input an &alpha; = 1, then the code will run DCGAN (see paper for proof).

# Evaluation code
The repository has `fid_varying_renyigan.py` and `fid_static_renyigan.py` codes,
which are evaluation codes use to compute the raw FID scores for RenyiGAN-[&beta;<sub>1</sub>, &beta;<sub>2</sub>] 
and RenyiGAN-&alpha; respectively.
To evaluate RenyiGAN-[&beta;<sub>1</sub>, &beta;<sub>2</sub>], simply run
```
python3 fid_varying_renyigan.py
``` 
which will prompt you to enter the version, subversion, and the trial number. 
To evaluate RenyiGAN-&alpha;, simply run
```
python3 fid_static_renyigan.py
``` 
which will prompt you to enter the alpha value and version.
Note that the evaluation scripts use on the numpy (`.npy`) files that the training files save, 
and delete all but the best performing epoch's generated images. 
We sampled 10,000 real and generated images to calculate the FID scores. 

# Further information 
We initialized the weights of each layer by using a Gaussian random variable with mean 0 and standard deviation
0.01.
We also used the Adam optimizer with a learning rate of 
 2 x 10<sup>-4</sup>, &beta;<sub>1</sub> = 0.5, &beta;<sub>2</sub> = 0.999, and
&epsilon; = 1 x 10<sup>-7</sup> for both networks. 
The batch size was chosen to be 100 for the 60,000 MNIST images.
The total number of epochs was 250 for the MNIST images. 

You can find sample images and plots for each version in `images/`. 


RenyiGan-TensorFlow2/History.txt
FIDScore: [1, 2, 1]
HELLO

renyigan_varying_alpha.py
import os
import time
import tensorflow as tf
from tensorflow.keras.initializers import RandomNormal
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense, BatchNormalization, \
    LeakyReLU, Conv2DTranspose, Conv2D, Dropout, Flatten, Reshape
import utils
import numpy as np

seed_num, trial_num, version, subversion = input("Seed, trial number, version, subversion: ").split()
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
np.random.seed(int(seed_num))
tf.random.set_random_seed(int(seed_num))
if int(subversion) == 3:
    alpha = 1.1
else:
    alpha = 0.0

class GAN(object):
    def __init__(self, alpha, trial_num, version, subversion):
        self.batch_size = 100
        self.n_classes = 10
        self.buffer_size = 50000
        self.training = True
        self.alpha = alpha
        self.alpha_placeholder = tf.placeholder(tf.float32, shape=[], name='alpha_placeholder')
        self.version = version
        self.subversion = subversion
        self.trial_num = trial_num
        self.noise_dim = 28 * 28
        self.dropout_constant = 0.6
        self.epsilon = 0.0001  # To ensure the log doesn't blow up to -infinity
        self.predictions = []
        self._make_directory('data')
        self._make_directory('data/varying')
        self._make_directory('data/varying/v' + str(self.version) + '-' + str(self.subversion))
        self._make_directory('data/varying/v' + str(self.version) + '-' + str(self.subversion) + '/trial' + str(self.trial_num))

    @staticmethod
    def _make_directory(PATH):
        if not os.path.exists(PATH):
            os.mkdir(PATH)

    def get_data(self):
        with tf.name_scope('data'):
            train_data, test_data = utils.get_mnist_dataset(self.batch_size)
            self.iterator = tf.data.Iterator.from_structure(train_data.output_types,
                                                            train_data.output_shapes)
            img, _ = self.iterator.get_next()
            self.img = tf.reshape(img, shape=[-1, 28, 28, 1])

            self.train_init = self.iterator.make_initializer(train_data)
            self.test_init = self.iterator.make_initializer(test_data)

    def build_generator(self):
        with tf.name_scope('generator') as scope:
            model = Sequential(name=scope)
            model.add(Dense(7 * 7 * 256, use_bias=False, kernel_initializer=
            RandomNormal(mean=0.0, stddev=0.01), input_shape=(self.noise_dim,)))
            model.add(BatchNormalization())
            model.add(LeakyReLU())

            model.add(Reshape((7, 7, 256)))
            assert model.output_shape == (None, 7, 7, 256)  # Note: None is the batch size

            model.add(Conv2DTranspose(128, (5, 5), strides=(1, 1), padding='same', use_bias=False, kernel_initializer=
            RandomNormal(mean=0.0, stddev=0.01)))
            assert model.output_shape == (None, 7, 7, 128)
            model.add(BatchNormalization())
            model.add(LeakyReLU())

            model.add(Conv2DTranspose(64, (5, 5), strides=(2, 2), padding='same', use_bias=False, kernel_initializer=
            RandomNormal(mean=0.0, stddev=0.01)))
            assert model.output_shape == (None, 14, 14, 64)
            model.add(BatchNormalization())
            model.add(LeakyReLU())

            model.add(Conv2DTranspose(1, (5, 5), strides=(2, 2), padding='same', activation='tanh', use_bias=False,
                                      kernel_initializer=RandomNormal(mean=0.0, stddev=0.01)))
            assert model.output_shape == (None, 28, 28, 1)

            return model

    def build_discriminator(self):
        with tf.name_scope('discriminator') as scope:
            model = Sequential(name=scope)
            model.add(Conv2D(64, (5, 5), strides=(2, 2), padding='same', kernel_initializer=
            RandomNormal(mean=0.0, stddev=0.01)))
            model.add(LeakyReLU())
            model.add(Dropout(0.3))

            model.add(Conv2D(128, (5, 5), strides=(2, 2), padding='same', kernel_initializer=
            RandomNormal(mean=0.0, stddev=0.01)))
            model.add(LeakyReLU())
            model.add(Dropout(0.3))

            model.add(Flatten())
            model.add(Dense(1, activation='sigmoid', kernel_initializer=
            RandomNormal(mean=0.0, stddev=0.01)))

            return model

    # Vanilla DCGAN discriminator loss function
    def dis_loss_vgan(self):
        with tf.name_scope('disLossVGAN'):
            real_loss = -tf.math.log(self.real_output + self.epsilon)
            real_loss = tf.math.reduce_mean(real_loss)

            fake_loss = -tf.math.log(1 - self.fake_output + self.epsilon)
            fake_loss = tf.math.reduce_mean(fake_loss)

            dis_loss = real_loss + fake_loss

            return dis_loss

    # Vanilla DCGAN discriminator loss function with gradient penalty
    def dis_loss_vgan_gp(self):
        with tf.name_scope('disLossVGANGP'):
            real_loss = -tf.math.log(self.real_output + self.epsilon)
            real_loss = tf.math.reduce_mean(real_loss)

            fake_loss = -tf.math.log(1 - self.fake_output + self.epsilon)
            fake_loss = tf.math.reduce_mean(fake_loss)

            gradients = tf.gradients(-tf.math.log(1 / self.real_output - 1), [self.img])[0]
            r1_penalty = tf.reduce_mean(tf.reduce_sum(tf.square(gradients), axis=[1, 2, 3]))
            dis_loss = real_loss + fake_loss + 5 * r1_penalty

            return dis_loss

    # RenyiGAN discriminator loss function
    def dis_loss(self):
        with tf.name_scope('disLoss'):
            real_loss = tf.math.reduce_mean(tf.math.pow(self.real_output, (self.alpha - 1)
                                                        * tf.ones_like(self.real_output)))
            real_loss = 1.0 / (self.alpha - 1) * tf.math.log(real_loss + self.epsilon) + tf.math.log(2.0)
            f = tf.math.reduce_mean(tf.math.pow(1 - self.fake_output,
                                                (self.alpha - 1) * tf.ones_like(self.fake_output)))
            gen_loss = 1.0 / (self.alpha - 1) * tf.math.log(f + self.epsilon) + tf.math.log(2.0)
            dis_loss = - real_loss - gen_loss
            return dis_loss

    # Vanilla DCGAN generator l1 loss function (mostly for testing)
    def gen_loss_vgan_l1(self):
        with tf.name_scope('genLossVGANL1'):
            fake_loss = tf.math.log(1 - self.fake_output + self.epsilon)
            fake_loss = tf.math.reduce_mean(fake_loss)
            gen_loss = tf.math.abs(fake_loss + tf.math.log(2.0))
            return gen_loss

    # Vanilla DCGAN generator loss function
    def gen_loss_vgan(self):
        print("DCGAN")
        with tf.name_scope('genLossVGAN'):
            fake_loss = tf.math.log(1 - self.fake_output + self.epsilon)
            gen_loss = tf.math.reduce_mean(fake_loss)
            return gen_loss

    # RenyiGAN generator loss function 
    def gen_loss(self):
        print("RenyiGAN " + str(self.alpha))
        with tf.name_scope('genLoss'):
            f = tf.math.reduce_mean(tf.math.pow(1 - self.fake_output, 
                (self.alpha_placeholder - 1) * tf.ones_like(self.fake_output)))
            gen_loss = 1.0 / (self.alpha_placeholder - 1) * tf.math.log(f + self.epsilon)
            return gen_loss

    # RenyiGAN generator loss function (has l1 norm incorporated)
    def gen_loss_l1(self):
        print("RenyiGAN " + str(self.alpha))
        with tf.name_scope('genLossL1'):
            f = tf.math.reduce_mean(tf.math.pow(1 - self.fake_output,
                                                (self.alpha_placeholder - 1) * tf.ones_like(self.fake_output)))
            gen_loss = tf.math.abs(
                1.0 / (self.alpha_placeholder - 1) * tf.math.log(f + self.epsilon) + tf.math.log(2.0))
            return gen_loss

    def optimize(self):
        self.gen_opt = tf.train.AdamOptimizer(2e-4, beta1=0.5, name="generator_optimizer")
        self.gen_opt_minimize = self.gen_opt.minimize(self.gen_loss_value, var_list=self.generator.trainable_variables)
        self.dis_opt = tf.train.AdamOptimizer(2e-4, beta1=0.5, name="discriminator_optimizer")
        self.dis_opt_minimize = self.dis_opt.minimize(self.dis_loss_value,
                                                      var_list=self.discriminator.trainable_variables)

    def gen_loss_wrapper(self):
        self.gen_loss_value = tf.cond(tf.cast(self.alpha_placeholder == 1.0, tf.bool), lambda: self.gen_loss_vgan(),
                                      lambda: self.gen_loss())

    def gen_loss_wrapper_l1(self):
        self.gen_loss_value = tf.cond(tf.cast(self.alpha_placeholder == 1.0, tf.bool), lambda: self.gen_loss_vgan_l1(),
                                      lambda: self.gen_loss_l1())

    def build(self):
        self.get_data()
        self.generator = self.build_generator()
        self.discriminator = self.build_discriminator()
        self.fake_output_images = self.generator(tf.random.normal([self.batch_size, self.noise_dim]))
        self.fake_output = self.discriminator(self.fake_output_images)
        self.real_output = self.discriminator(self.img)
        if self.version == 1 or self.version == 3:
            self.gen_loss_wrapper()
        else:
            self.gen_loss_wrapper_l1()
        if self.version == 1 or self.version == 2:
            self.dis_loss_value = self.dis_loss_vgan()
        else:
            self.dis_loss_value = self.dis_loss_vgan_gp()
        self.optimize()

    def train_one_epoch(self, sess, init, epoch):
        start_time = time.time()
        sess.run(init)
        self.training = True
        total_loss_gen = 0
        total_loss_dis = 0
        n_batches = 0
        try:
            while True:
                _, disLoss = sess.run([self.dis_opt_minimize, self.dis_loss_value])
                _, genLoss = sess.run([self.gen_opt_minimize, self.gen_loss_value],
                                      feed_dict={self.alpha_placeholder: self.alpha})
                total_loss_gen += genLoss
                total_loss_dis += disLoss
                n_batches += 1
        except tf.errors.OutOfRangeError:
            pass
        self.save_generated_images(sess, epoch)
        print('Average generator loss at epoch {0}: {1}'.format(epoch, total_loss_gen / n_batches))
        print('Average discriminator loss at epoch {0}: {1}'.format(epoch, total_loss_dis / n_batches))
        print('Took: {0} seconds'.format(time.time() - start_time))

    def save_generated_images(self, sess, epoch):
        temp = self.generator(tf.random.normal([self.buffer_size, self.noise_dim]))
        temp = sess.run(temp)
        if len(self.predictions) > 0:
            self.predictions.pop(0)
        self.predictions.append(temp)
        np.save('data/varying/v' + str(self.version) + '-' + str(self.subversion) 
                + '/trial' + str(self.trial_num) + '/predictions' + str(epoch),
                self.predictions)

    def train(self, n_epochs):
        self._make_directory('checkpoints')
        self._make_directory('checkpoints/varying/')
        self._make_directory('checkpoints/varying/v' + str(self.version) + '-' + str(self.subversion))
        self.cpt_PATH = 'checkpoints/varying/v' + str(self.version) + '-' + str(self.subversion) + '/trial' + str(self.trial_num)
        if self.trial_num == 1:
            self._make_directory(self.cpt_PATH)

        go_up = True
        with tf.Session() as sess:
            sess.run(tf.global_variables_initializer())
            sess.run(self.train_init)  # run iterator initilizer to save properly
            checkpoint = tf.train.Saver(
                {'generator_optimizer': self.gen_opt, 'discriminator_optimizer': self.dis_opt,
                 'generator': self.generator, 'discriminator': self.discriminator, 'iterator': self.iterator},
                max_to_keep=3)
            for epoch in range(n_epochs):
                if self.trial_num == 1:
                    if epoch % 10 == 0:
                        save_path = checkpoint.save(sess, self.cpt_PATH + str('/ckpt'), global_step=epoch)
                        print("Saved checkpoint for step {}: {}".format(int(epoch), save_path))
                print("Alpha value: " + str(self.alpha))
                self.train_one_epoch(sess, self.train_init, epoch)
                # Subversion 1 between [0, 0.9]
                if self.subversion == 1:
                    if go_up:
                        self.alpha = round(self.alpha + 0.1, 1)
                        if self.alpha == 0.9:
                            go_up = False
                    else:
                        self.alpha = round(self.alpha - 0.1, 1)
                        if self.alpha == 0.0:
                            go_up = True
                # Subversion 2 between [0, 3]
                elif self.subversion == 2:
                    if go_up:
                        self.alpha = round(self.alpha + 0.1, 1)
                        if self.alpha == 3.0:
                            go_up = False
                        if self.alpha == 1.0:
                            self.alpha = round(1.1, 1)
                    else:
                        self.alpha = round(self.alpha - 0.1, 1)
                        if self.alpha == 0.0:
                            go_up = True
                        if self.alpha == 1.0:
                            self.alpha = round(0.9, 1)
                # Subversion 3 between [1.1, 4]
                else:
                    if go_up:
                        self.alpha = round(self.alpha + 0.1, 1)
                        if self.alpha == 4.0:
                            go_up = False
                    else:
                        self.alpha = round(self.alpha - 0.1, 1)
                        if self.alpha == 1.1:
                            go_up = True


model = GAN(round(float(alpha), 1), int(trial_num), int(version), int(subversion))
model.build()
model.train(n_epochs=250)


